File: credit-memo-total-not-refreshed/python/flag_creditmemo_total_drift.py
"""Flag Magento 2 credit memos whose grand_total was never refreshed after an
adjustment edit.

In the admin credit memo creation form, the grand total shown and saved is
only recalculated by the Update Qty's JavaScript handler, which fires on item
quantity changes. It is never wired to the Refund Shipping, Adjustment Refund
(adjustment_positive), or Adjustment Fee (adjustment_negative) input fields,
so editing those alone can leave grand_total stale in both the UI and the
persisted record unless a qty update or the actual submission forces
Magento's server side total collectors to run. The same drift is reachable
through POST /V1/creditmemo, since the API does not independently re-validate
the total. There is no supported endpoint to fix a posted creditmemo's
grand_total, so this only reports the drift. Run on a schedule. Safe to run
again and again.
"""
import os
import logging
import datetime
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_creditmemo_total_drift")

MAGENTO_URL = os.environ["MAGENTO_URL"].rstrip("/")
TOKEN = os.environ["MAGENTO_ADMIN_TOKEN"]
LOOKBACK_DAYS = float(os.environ.get("LOOKBACK_DAYS", "7"))
EPSILON = float(os.environ.get("EPSILON", "0.01"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def magento_get(path, params=None):
    r = requests.get(
        f"{MAGENTO_URL}/rest/V1{path}",
        params=params or {},
        headers={"Authorization": f"Bearer {TOKEN}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def since_iso(lookback_days):
    since = datetime.datetime.utcnow() - datetime.timedelta(days=lookback_days)
    return since.strftime("%Y-%m-%d %H:%M:%S")


def recent_creditmemos(since, page_size=100, current_page=1):
    params = {
        "searchCriteria[filterGroups][0][filters][0][field]": "created_at",
        "searchCriteria[filterGroups][0][filters][0][conditionType]": "gteq",
        "searchCriteria[filterGroups][0][filters][0][value]": since,
        "searchCriteria[pageSize]": page_size,
        "searchCriteria[currentPage]": current_page,
    }
    return magento_get("/creditmemos", params)["items"]


def normalize_creditmemo(raw):
    return {
        "entityId": raw.get("entity_id"),
        "incrementId": raw.get("increment_id"),
        "orderId": raw.get("order_id"),
        "subtotal": float(raw.get("subtotal") or 0),
        "discountAmount": float(raw.get("discount_amount") or 0),
        "shippingAmount": float(raw.get("shipping_amount") or 0),
        "taxAmount": float(raw.get("tax_amount") or 0),
        "adjustmentPositive": float(raw.get("adjustment_positive") or 0),
        "adjustmentNegative": float(raw.get("adjustment_negative") or 0),
        "grandTotal": float(raw.get("grand_total") or 0),
    }


def evaluate_creditmemo_total_drift(creditmemo, epsilon=0.01):
    expected_grand_total = round(
        creditmemo["subtotal"]
        - creditmemo["discountAmount"]
        + creditmemo["shippingAmount"]
        + creditmemo["taxAmount"]
        + creditmemo["adjustmentPositive"]
        - creditmemo["adjustmentNegative"],
        2,
    )
    delta = round(creditmemo["grandTotal"] - expected_grand_total, 2)
    is_drifted = abs(delta) > epsilon
    return {
        "expectedGrandTotal": expected_grand_total,
        "delta": delta,
        "isDrifted": is_drifted,
    }


def run():
    since = since_iso(LOOKBACK_DAYS)
    flagged = []
    page = 1
    while True:
        raw_items = recent_creditmemos(since, current_page=page)
        if not raw_items:
            break
        for raw in raw_items:
            creditmemo = normalize_creditmemo(raw)
            result = evaluate_creditmemo_total_drift(creditmemo, EPSILON)
            if result["isDrifted"]:
                flagged.append({**creditmemo, **result})
        if len(raw_items) < 100:
            break
        page += 1

    for row in flagged:
        log.warning(
            "Creditmemo %s (order %s) grand_total drifted. stored=%.2f expected=%.2f delta=%.2f",
            row["incrementId"], row["orderId"], row["grandTotal"], row["expectedGrandTotal"], row["delta"],
        )

    if flagged:
        log.error("%d credit memo(s) drifted. This script never edits them directly.", len(flagged))
    else:
        log.info("Done. No credit memo total drift found.")


if __name__ == "__main__":
    run()

File: credit-memo-total-not-refreshed/node/flag-creditmemo-total-drift.js
/**
 * Flag Magento 2 credit memos whose grand_total was never refreshed after an
 * adjustment edit.
 *
 * In the admin credit memo creation form, the grand total shown and saved is
 * only recalculated by the Update Qty's JavaScript handler, which fires on
 * item quantity changes. It is never wired to the Refund Shipping,
 * Adjustment Refund (adjustment_positive), or Adjustment Fee
 * (adjustment_negative) input fields, so editing those alone can leave
 * grand_total stale in both the UI and the persisted record unless a qty
 * update or the actual submission forces Magento's server side total
 * collectors to run. The same drift is reachable through POST
 * /V1/creditmemo, since the API does not independently re-validate the
 * total. There is no supported endpoint to fix a posted creditmemo's
 * grand_total, so this only reports the drift. Run on a schedule. Safe to
 * run again and again.
 *
 * Guide: https://www.allanninal.dev/magento/credit-memo-total-not-refreshed/
 */
import { pathToFileURL } from "node:url";

const MAGENTO_URL = (process.env.MAGENTO_URL || "https://demo.example.com").replace(/\/$/, "");
const TOKEN = process.env.MAGENTO_ADMIN_TOKEN || "token_dummy";
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 7);
const EPSILON = Number(process.env.EPSILON || 0.01);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function evaluateCreditmemoTotalDrift(creditmemo, epsilon = 0.01) {
  const expectedGrandTotal = round2(
    creditmemo.subtotal
    - creditmemo.discountAmount
    + creditmemo.shippingAmount
    + creditmemo.taxAmount
    + creditmemo.adjustmentPositive
    - creditmemo.adjustmentNegative
  );
  const delta = round2(creditmemo.grandTotal - expectedGrandTotal);
  const isDrifted = Math.abs(delta) > epsilon;
  return { expectedGrandTotal, delta, isDrifted };
}

function round2(n) {
  return Math.round(n * 100) / 100;
}

async function magentoGet(path, params = {}) {
  const url = new URL(`${MAGENTO_URL}/rest/V1${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${TOKEN}` } });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

function sinceIso(lookbackDays) {
  const since = new Date(Date.now() - lookbackDays * 86400 * 1000);
  return since.toISOString().slice(0, 19).replace("T", " ");
}

async function recentCreditmemos(since, pageSize = 100, currentPage = 1) {
  const params = {
    "searchCriteria[filterGroups][0][filters][0][field]": "created_at",
    "searchCriteria[filterGroups][0][filters][0][conditionType]": "gteq",
    "searchCriteria[filterGroups][0][filters][0][value]": since,
    "searchCriteria[pageSize]": pageSize,
    "searchCriteria[currentPage]": currentPage,
  };
  const data = await magentoGet("/creditmemos", params);
  return data.items;
}

function normalizeCreditmemo(raw) {
  return {
    entityId: raw.entity_id,
    incrementId: raw.increment_id,
    orderId: raw.order_id,
    subtotal: Number(raw.subtotal || 0),
    discountAmount: Number(raw.discount_amount || 0),
    shippingAmount: Number(raw.shipping_amount || 0),
    taxAmount: Number(raw.tax_amount || 0),
    adjustmentPositive: Number(raw.adjustment_positive || 0),
    adjustmentNegative: Number(raw.adjustment_negative || 0),
    grandTotal: Number(raw.grand_total || 0),
  };
}

export async function run() {
  const since = sinceIso(LOOKBACK_DAYS);
  const flagged = [];
  let page = 1;

  while (true) {
    const rawItems = await recentCreditmemos(since, 100, page);
    if (!rawItems.length) break;

    for (const raw of rawItems) {
      const creditmemo = normalizeCreditmemo(raw);
      const result = evaluateCreditmemoTotalDrift(creditmemo, EPSILON);
      if (result.isDrifted) flagged.push({ ...creditmemo, ...result });
    }

    if (rawItems.length < 100) break;
    page++;
  }

  for (const row of flagged) {
    console.warn(
      `Creditmemo ${row.incrementId} (order ${row.orderId}) grand_total drifted. ` +
      `stored=${row.grandTotal.toFixed(2)} expected=${row.expectedGrandTotal.toFixed(2)} delta=${row.delta.toFixed(2)}`
    );
  }

  if (flagged.length) {
    console.error(`${flagged.length} credit memo(s) drifted. This script never edits them directly.`);
  } else {
    console.log("Done. No credit memo total drift found.");
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
